### ISSUE TITLE
Decide how to implement difficulty selection

### ISSUE THREAD

**Issue by ldicker83** [OutpostUniverse/OPHD#200]
So I see that @cugone is beginning to work on the difficulty implementation. Thank you for your efforts! It's very much appreciated especially while I'm recovering from illness.

---

So the point of this topic is to decide exactly how to select difficulty. The 'planet select' screen is sort of a stub screen at this point. The difficulty selection that's been presented is very much like the original game, something I'd hoped to avoid.

What I'm envisioning is a game setup screen that present all of these options as a sort of full-screen UI. I'll finish fleshing this out later today...

**Comment by DanRStevens** [comment]
Yeah, I'm finding the new popup slows down the process of creating a new game.

I think it might flow smoother if the difficult selection and the planet selection were done on the same screen. Potentially you could skip selecting a difficulty setting, and just have it use a default value, making the selection optional. That way you still only need one click (on the planet type) to start a new game.

I suppose the initial default should be "easy". If the player changes the difficulty, perhaps it can be remembered for future games.

As for display layout, maybe place a horizontal list of difficulty settings above the list of planets. Since clicking the planet starts the game, it should be below any optional settings.

**Comment by cugone** [comment]
I agree, moving the difficulty options to the `PlanetSelectionState` would be better.

The differences between each difficulty aren't totally explained even in the original. The only thing I've been able to determine is the frequency of disasters. Beginner having none and each difficulty afterwards increasing the frequency.

**Comment by DanRStevens** [comment]
Maybe we can eventually add some sort of mouse-over popup that summarizes what the difficulty setting changes.

**Comment by ldicker83** [comment]
I thought I posted this already but anyway, here goes.

The idea here is to make this as visual as possible with plenty of details, the sort of thing that modern games do. Attached is an archive with a very basic visual layout that I'm considering. Note that color scheme isn't set in stone -- I've been using the black/green as a sort of nod to the original (see the fullscreen UI's) but it's definitely not set in stone.

I didn't note what the images are exactly but basically the first set is some sort of icon that indicates visually a difficulty level (I'm no artist so I leave that up to someone better suited to that to offer ideas), the next set I think is for just the planets? I'm not sure a visual for stars is really necessary.

The final one shows the planet selection rotating.

I just realized now that I forgot about a site map choice but ultimately I'd like to offer a number of site maps for planets (minimum of one, maximum of 10) with the user being able to expand into multiple sites and create routes between them. That's a later part of development though but assume that after planet choice the user then gets to choose a site map for landing.

[game-setup-demo.zip](https://github.com/OutpostUniverse/OPHD/files/4209367/game-setup-demo.zip)

**Comment by ldicker83** [comment]
Side note -- difficulty for OPHD will affect:

- Initial population count (what comes down from the ship)
- Initial resource count (what comes down from the ship)
- Disaster frequency
- Intensity of disasters
- Choice of stars
- Choice of Planets
- Mine Counts
- Mine Yields
- Population growth
- Effect of morale on factory production and research
- Decreased positive morale effects and increased negative morale effects
- How critically the population responds to your decisions (related to the above)
- Structure costs
- Recycling values of razed structures and landers
- Research times
- Benefits of research (e.g., if base beneficial effect of a completed research topic is 10% reduction in cost of structures for normal difficulty, beginner would be 12%, normal 10%, hard 7.5% and impossible 5%... arbitrary numbers but illustrates the thinking here)

Probably a few other factors but basically difficulty here can be potentially punishing especially on 'nightmare mode'.

The current state of the game is very close to how I want 'difficult' to be in terms of resource production. I wanted to get a sort of balance right for 'hard' that's not impossible but is also kinda hard to get right. Normal modes would offer more initial resources

**Comment by DanRStevens** [comment]
Direct posting the images from the zip file Leeor attached for ease of reference:

![difficulty_select](https://user-images.githubusercontent.com/16793474/78045669-76b01900-7333-11ea-8f63-3e2e82f3806f.png)
![star_select](https://user-images.githubusercontent.com/16793474/78045674-7879dc80-7333-11ea-91bc-a48b61d44487.png)
![planet_select_1](https://user-images.githubusercontent.com/16793474/78045672-77e14600-7333-11ea-8586-2299cd94748e.png)
![planet_select_2](https://user-images.githubusercontent.com/16793474/78045673-77e14600-7333-11ea-9380-6ed7558a561b.png)
![finalize](https://user-images.githubusercontent.com/16793474/78045671-7748af80-7333-11ea-9005-f6b2fb953818.png)

**Comment by DanRStevens** [comment]
Looks like the "Star Select" and the second "Planet Select" screens are identical. I assume only one of those was intended.

----

To add my thoughts to that, it might be nice if we could reduce clicking. Instead of having next buttons, clicking large images representing the difficulty or planet could proceed to the next screen. Details information about the options can be displayed on mouse over. Having a back button is fine. The "Star Select" screen might need to be reworked to support such a flow, as none of the options seem to represent a final proceed to the next screen type of selection.

For the "Star Select" screen, it seems there are a few options. Maybe that one can be re-worked to be a bit more graphical. In particular, the radio buttons can be replaced by sequences of clickable images. Same for the check box, which can be replaced by two images, and function much like the radio buttons converted to graphical form.

----

Actually, I wonder if it would be possible to present everything on one screen. That would eliminate the forward/back navigation. Instead, if space is limited, we might have various areas enlarge on mouse over, so we can see more detail about a given set of parameters.

I imagine the three sections to be stacked, with various toggle icons for the options. Initially the difficulty select and star select options could be small, and represent default options, while the planet select is large. If the user moves the mouse over the difficulty select section, that area could enlarge, while the previously enlarged section shrinks a bit. That would help draw the user's attention to a given area. If the mouse is not over any particular area, I think it should revert to making the planet select area large. That would be the section where making a selection would close the screen and proceed to start the game.

**Comment by cugone** [comment]
I assume "Star Select" is similar to the original game. You select a star system first (not knowing what planets are available), then a planet within it.

**Comment by DanRStevens** [comment]
Hmm, that creates some interesting opportunities for animations. When changing the star, we may need to change the planets. Rather than just have them abruptly change, perhaps we can animate the change. Maybe have the planets fly off, and new ones fly in. Or perhaps the planets can be illustrated as some kind of solar system, and changing the star make the view fly to another solar system.

I think the details about each selection should all appear on mouse over.

**Comment by cugone** [comment]
It depends on the amount of text. For small amounts, sure show it on hover. But for large amounts you want it to stay up when it's clicked so the user can read it.

**Comment by DanRStevens** [comment]
To be clear, I'm still talking about having a permanent display area for the text, not a popup tooltip text display. I was thinking it should work the same as the current planet select screen in OPHD. On mouse over, the planet spins, and descriptive text is displayed at the bottom of the screen.

**Comment by ldicker83** [comment]
The star select screen isn't too much like the original game -- it doesn't have an effect on the availability of planets but it does affect other things like resource availability, light output, stellar storm frequency, etc.

The 'text area' at the bottom of the screens is intended to be that 'hover text' area -- there won't be a box drawn around it, that's just there to visualize the layout of the components.

As a note, I'm entirely open to suggestions here especially in terms of Stellar choice and exactly what variables star choice will have on the simulation.

**Comment by ldicker83** [comment]
> Hmm, that creates some interesting opportunities for animations. When changing the star, we may need to change the planets. Rather than just have them abruptly change, perhaps we can animate the change. Maybe have the planets fly off, and new ones fly in. Or perhaps the planets can be illustrated as some kind of solar system, and changing the star make the view fly to another solar system.

That's the general idea. I have a tweening engine I was experimenting with way back when. Would ultimately be part of the polishing stage but yeah, having things slide around and animated would be slick :D

**Comment by ldicker83** [comment]
Reread some of @DanRStevens's comments and wanted to go over some of it.

> To add my thoughts to that, it might be nice if we could reduce clicking. Instead of having next buttons, clicking large images representing the difficulty or planet could proceed to the next screen. Details information about the options can be displayed on mouse over. Having a back button is fine. The "Star Select" screen might need to be reworked to support such a flow, as none of the options seem to represent a final proceed to the next screen type of selection.

I generally like this idea. It's better than what originally put together.

> For the "Star Select" screen, it seems there are a few options. Maybe that one can be re-worked to be a bit more graphical. In particular, the radio buttons can be replaced by sequences of clickable images. Same for the check box, which can be replaced by two images, and function much like the radio buttons converted to graphical form.

I like this idea, though I would want to be sure that it's as intuitive as possible without having to have too much layer stacking.

> Actually, I wonder if it would be possible to present everything on one screen. That would eliminate the forward/back navigation. Instead, if space is limited, we might have various areas enlarge on mouse over, so we can see more detail about a given set of parameters.

That could be done using 'panels' that expand when clicked on, almost like clicking on a card to see what's on it. I wouldn't want everything to be all on one screen at once as it can be really intimidating and I imagine the screen could get very busy. I like the idea of modern EX designs where smaller chunks of information are presented to the user at a time.

> I imagine the three sections to be stacked, with various toggle icons for the options. Initially the difficulty select and star select options could be small, and represent default options, while the planet select is large. If the user moves the mouse over the difficulty select section, that area could enlarge, while the previously enlarged section shrinks a bit. That would help draw the user's attention to a given area. If the mouse is not over any particular area, I think it should revert to making the planet select area large. That would be the section where making a selection would close the screen and proceed to start the game.

This sounds a lot like my card metaphor. Could expand on this one a bit.
